modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C

```diff
@@ -0,0 +1,443 @@
+{
+  std::stringstream iter_output;
+  std::stringstream * iter_output_ptr = (_output_iteration_info ? &iter_output : nullptr);
+
+  if (!_legacy_residual)
+  {
+    if (!internalSolve(effective_trial_stress, scalar, iter_output_ptr))
+    {
+      if (iter_output_ptr)
+        mooseError(iter_output_ptr->str());
+      else
+      {
+        internalSolve(effective_trial_stress, scalar, &iter_output);
+        mooseError(iter_output.str());
+      }
+    }
+    else if (iter_output_ptr)
+      console << iter_output_ptr->str();
+  }
+  else
+  {
+    if (!internalSolveLegacy(effective_trial_stress, scalar, iter_output_ptr))
+    {
+      if (iter_output_ptr)
+        mooseError(iter_output_ptr->str());
+      else
+      {
+        internalSolveLegacy(effective_trial_stress, scalar, &iter_output);
+        mooseError(iter_output.str());
+      }
+    }
+    else if (iter_output_ptr)
+      console << iter_output_ptr->str();
+  }
+}
+
+bool
+NewtonIterationSingleVariable::internalSolve(const Real effective_trial_stress,
+                                             Real & scalar,
+                                             std::stringstream * iter_output)
+{
+  scalar = 0.0;
+  Real scalar_old = 0.0;
+  Real scalar_increment = 0.0;
+  const Real min_permissible_scalar = 0.0;
+  const Real max_permissible_scalar = maximumPermissibleValue(effective_trial_stress);
+  Real scalar_upper_bound = max_permissible_scalar;
+  Real scalar_lower_bound = 0.0;
+  unsigned int it = 0;
+  const Real dt = getDT();
+
+  if (effective_trial_stress == 0.0)
+  {
+    outputIterInfo(iter_output, it, effective_trial_stress, scalar, 0.0, 1.0);
+    return true;
+  }
+
+  Real residual = computeResidualInternal(effective_trial_stress, scalar, dt);
+  Real residual_old = residual;
+
+  Real init_resid_norm = std::abs(residual);
+  if (init_resid_norm == 0.0)
+    init_resid_norm = 1.0;
+  Real init_resid_sign = (residual < 0.0 ? -1.0 : 1.0);
+
+  Real reference_residual = computeReferenceResidual(effective_trial_stress, scalar);
+
+  if (converged(residual, reference_residual))
+  {
+    iterationFinalize(scalar);
+    outputIterInfo(iter_output, it, effective_trial_stress, scalar, residual, reference_residual);
+    return true;
+  }
+
+  while (it < _max_its && !converged(residual, reference_residual))
+  {
+    scalar_increment = -residual / computeDerivativeInternal(effective_trial_stress, scalar, dt);
+    scalar = scalar_old + scalar_increment;
+
+    if (_check_range)
+      checkPermissibleRange(scalar,
+                            scalar_increment,
+                            scalar_old,
+                            min_permissible_scalar,
+                            max_permissible_scalar,
+                            iter_output);
+
+    residual = computeResidualInternal(effective_trial_stress, scalar, dt);
+    reference_residual = computeReferenceResidual(effective_trial_stress, scalar);
```

Comment:
Similar idea.  It's the quantity that we compare the residual to for relative convergence.